src/manifest.ts

```diff
@@ -1114,28 +1133,145 @@ export class Manifest {
+        for (const pullNumber in releasesByPullRequest) {
+          const releases = await this.createReleasesForPullRequest(
             releasesByPullRequest[pullNumber],
             pullRequestsByNumber[pullNumber]
-          )
-        );
+          );
+          resultReleases.push(...releases);
+        }
+        return resultReleases;
+      } else {
+        const promises: Promise<CreatedRelease[]>[] = [];
+        for (const pullNumber in releasesByPullRequest) {
+          promises.push(
+            this.createReleasesForPullRequest(
+              releasesByPullRequest[pullNumber],
+              pullRequestsByNumber[pullNumber]
+            )
+          );
+        }
+        const releases = await Promise.all(promises);
+        return releases.reduce((collection, r) => collection.concat(r), []);
+      }
+    };
+
+    // to minimize risks of race condition we attempt to lock branches used as the source of commits for the duration of the
+    // release process
+    let createdReleases: CreatedRelease[];
+    const pullRequests = Object.values(pullRequestsByNumber);
+    try {
+      const lockedBranches = await this.lockPullRequestsChangesBranch(
+        pullRequests
+      );
+      try {
+        // now that branches have been locked, ensure no new commits have been pushed since we created the release branch
+        await this.throwIfChangesBranchesRaceConditionDetected(pullRequests);
+        createdReleases = await runReleaseProcess();
+      } finally {
+        // always try to unlock branches, we don't want to keep them blocked as read-only when the release process fails
+        // before completion
+        await this.unlockPullRequestsChangesBranch(lockedBranches);
+      }
+    } catch (err: unknown) {
+      // Error: 403 "Resource not accessible by integration" is returned by GitHub when the API token doesn't have
+      // permissions to manipulate branch protection rules, if that seems to be the case we instead fallback to checking
+      // twice for race conditions, once at the beginning of the release process and once again before aligning branches.
+      // While not ideal that still significantly reduces risks of overwriting new commits.
+      //
+      // Error mentioned here: https://docs.github.com/en/code-security/code-scanning/troubleshooting-code-scanning/resource-not-accessible-by-integration
+      if (err && err instanceof RequestError && err.status === 403) {
+        await this.throwIfChangesBranchesRaceConditionDetected(pullRequests);
+        createdReleases = await runReleaseProcess();
+        await this.throwIfChangesBranchesRaceConditionDetected(pullRequests);
+      } else {
+        throw err;
       }
-      const releases = await Promise.all(promises);
-      return releases.reduce((collection, r) => collection.concat(r), []);
+    }
+
+    await this.alignPullRequestsChangesBranch(pullRequests);
+    return createdReleases;
+  }
```

Comment:
@marcel-stainless I may be over-thinking this but I wasn't too satisfied with my approach due to a detail: the `GITHUB_TOKEN` used in GitHub actions can only be given [a limited set of permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token), and the list doesn't include administration endpoints such as branch protection rules. That means branch locking can only be done if a Personal Access Token with admin permissions is passed to `release-please` instead of the default one from the github action context. That's not necessarily an issue, we likely can require that users provide an access tokens with admin permissions on the repository, but I don't know if that's something that should be forced upon Stainless customers.

I updated the implementation to first try the lock approach as it is the ideal scenario, and fallback to a double race condition check if branch permissions rules cannot be accessed.

I don't see this as blocking, I will merge the PR in its current state and we can continue to tweak things in future PRs.